### PR TITLE
Move Prompt to core libraries

### PR DESCRIPTION
**Pull request by donker**
This PR copies the Prompt classes and interfaces that are now scattered in the AE Library and Extensions projects to the DNN Core libraries. The goals are:

- to make the ability to create console commands more readily available to developers
- to clean up the code and make the coding experience for developers better and more intuitive

The new and the old system will run side by side until we deprecate the old system.

### CODE REVIEW

File: DNN Platform/Admin Modules/Dnn.Modules.Console/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -12,10 +12,5 @@
 [assembly: AssemblyTitle("Dnn.Modules.Console")]
 [assembly: AssemblyDescription("")]
 
-// Setting ComVisible to false makes the types in this assembly not visible 
-// to COM components.  If you need to access a type in this assembly from 
-// COM, set the ComVisible attribute to true on that type.
-[assembly: ComVisible(false)]
```
Comment (mitchelsellers) [inline DNN Platform/Admin Modules/Dnn.Modules.Console/Properties/AssemblyInfo.cs:18]: I don't believe this should have changed

File: DNN Platform/Admin Modules/Dnn.Modules.ModuleCreator/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -12,10 +12,5 @@
 [assembly: AssemblyTitle("Dnn.Modules.ModuleCreator")]
 [assembly: AssemblyDescription("")]
 
-// Setting ComVisible to false makes the types in this assembly not visible
```
Comment (mitchelsellers) [inline DNN Platform/Admin Modules/Dnn.Modules.ModuleCreator/Properties/AssemblyInfo.cs:15]: Same here

File: DNN Platform/Dnn.AuthServices.Jwt/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -18,11 +18,6 @@
 [assembly: AssemblyTrademark("")]
 [assembly: AssemblyCulture("")]
 
-// Setting ComVisible to false makes the types in this assembly not visible 
-// to COM components.  If you need to access a type in this assembly from 
-// COM, set the ComVisible attribute to true on that type.
-[assembly: ComVisible(false)]
```
Comment (mitchelsellers) [inline DNN Platform/Dnn.AuthServices.Jwt/Properties/AssemblyInfo.cs:24]: Same note

File: DNN Platform/Controls/CountryListBox/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -14,7 +14,6 @@
 [assembly: AssemblyTitle("DotNetNuke")]
 [assembly: AssemblyDescription("Open Source Web Application Framework")]
 
-[assembly: CLSCompliant(true)]
```
Comment (mitchelsellers) [inline DNN Platform/Controls/CountryListBox/Properties/AssemblyInfo.cs:17]: I would prefer we do this as a Separate PR if we want to disable this.

File: DNN Platform/Connectors/Azure/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -2,11 +2,10 @@
...
 // 
-﻿#region Usings
+#region Usings
 using System;
 using System.Reflection;
 #endregion
 
 [assembly: AssemblyTitle("DotNetNuke")]
 
-[assembly: CLSCompliant(true)]
```
Comment (mitchelsellers) [inline DNN Platform/Connectors/Azure/Properties/AssemblyInfo.cs:12]: I would prefer we do this as a Separate PR if we want to disable this.

File: DNN Platform/DotNetNuke.Abstractions/Portals/IPortalAliasInfo.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Portals/IPortalAliasInfo.cs:2]: This header is not correct.  Proper header format is

// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

Per guidance here: https://github.com/dotnet-foundation/foundation/blob/master/guidance/copyright.md

File: DNN Platform/DotNetNuke.Abstractions/Portals/IPortalSettings.cs [hunk 7]
```diff
@@ -2,18 +2,121 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT License. See LICENSE file in the project root for full license information.
 // 
-// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Portals/IPortalSettings.cs:6]: This change is incorrect as this is the proper license attribute.  The lines above this should be removed

File: DNN Platform/DotNetNuke.Abstractions/Prompt/ICommand.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE file in the project root for full license information.
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Prompt/ICommand.cs:3]: Incorrect copyright

File: DNN Platform/DotNetNuke.Abstractions/Prompt/ICommandHelp.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE file in the project root for full license information.
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Prompt/ICommandHelp.cs:3]: Another header just making note so we don't forget

File: DNN Platform/DotNetNuke.Abstractions/Prompt/ICommandOption.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Prompt/ICommandOption.cs:2]: Another header just making note so we don't forget

File: DNN Platform/DotNetNuke.Abstractions/Prompt/ICommandOption.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+// Licensed under the MIT License. See LICENSE file in the project root for full license information.
+// 
+namespace DotNetNuke.Abstractions.Prompt
+{
+    /// <summary>
+    /// This is used in the ICommandHelp to send a list of command parameters to the client for explanatory help
+    /// </summary>
+    public interface ICommandOption
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Prompt/ICommandOption.cs:10]: Not sure of the naming here.  This is named ICommandOption, but everywhere inside this object it is referred to itself as a Parameter?

File: DNN Platform/DotNetNuke.Abstractions/Prompt/ICommandRepository.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE file in the project root for full license information.
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Prompt/ICommandRepository.cs:3]: Another header just making note so we don't forget

File: DNN Platform/DotNetNuke.Abstractions/Prompt/IConsoleCommand.cs [hunk 13]
```diff
@@ -0,0 +1,46 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Prompt/IConsoleCommand.cs:2]: Another header just making note so we don't forget

File: DNN Platform/DotNetNuke.Abstractions/Prompt/IConsoleResultModel.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Prompt/IConsoleResultModel.cs:2]: Another header just making note so we don't forget

File: DNN Platform/DotNetNuke.Abstractions/Prompt/IPagingInfo.cs [hunk 15]
```diff
@@ -0,0 +1,30 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Prompt/IPagingInfo.cs:2]: Another header just making note so we don't forget

File: DNN Platform/DotNetNuke.Abstractions/Users/IUserInfo.cs [hunk 16]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Users/IUserInfo.cs:1]: This file is missing a header

File: DNN Platform/DotNetNuke.Abstractions/Users/IUserInfo.cs [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace DotNetNuke.Abstractions.Users
+{
+    public interface IUserInfo
+    {
+        int AffiliateID { get; set; }
+        //CacheLevel Cacheability { get; }
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Abstractions/Users/IUserInfo.cs:10]: If this isn't used let's remove it, same with other ones that are commented out

File: DNN Platform/DotNetNuke.Instrumentation/Properties/AssemblyInfo.cs [hunk 18]
```diff
@@ -11,12 +11,6 @@
 [assembly: AssemblyTitle("DotNetNuke.Instrumentation")]
 [assembly: AssemblyDescription("DotNetNuke's Instrumentation Framework")]
 
-// Setting ComVisible to false makes the types in this assembly not visible 
-// to COM components.  If you need to access a type in this assembly from 
-// COM, set the ComVisible attribute to true on that type.
-
-[assembly: ComVisible(false)]
```
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Instrumentation/Properties/AssemblyInfo.cs:18]: Same as others

File: DNN Platform/Library/Entities/Modules/Prompt/AddModule.cs [hunk 19]
```diff
@@ -0,0 +1,142 @@
+﻿using DotNetNuke.Abstractions.Portals;
```
Comment (mitchelsellers) [inline DNN Platform/Library/Entities/Modules/Prompt/AddModule.cs:1]: This file is missing the header

File: DNN Platform/Library/Entities/Modules/Prompt/ListModules.cs [hunk 20]
```diff
@@ -0,0 +1,111 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (mitchelsellers) [inline DNN Platform/Library/Entities/Modules/Prompt/ListModules.cs:2]: Another header just making note so we don't forget

File: DNN Platform/Library/Entities/Modules/Prompt/ListModules.cs [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+using DotNetNuke.Prompt;
+using DotNetNuke.Security.Permissions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DotNetNuke.Entities.Modules.Prompt
+{
+    /// <summary>
+    /// This is a (Prompt) Console Command. You should not reference this class directly. It is to be used solely through Prompt.
```
Comment (mitchelsellers) [inline DNN Platform/Library/Entities/Modules/Prompt/ListModules.cs:18]: This summary seems to be incorrect

File: DNN Platform/Library/Entities/Modules/Prompt/PromptModuleInfo.cs [hunk 22]
```diff
@@ -0,0 +1,52 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (mitchelsellers) [inline DNN Platform/Library/Entities/Modules/Prompt/PromptModuleInfo.cs:2]: Another header just making note so we don't forget

File: DNN Platform/Library/Entities/Modules/Settings/SerializationController.cs [hunk 23]
```diff
@@ -0,0 +1,194 @@
+﻿// 
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (mitchelsellers) [inline DNN Platform/Library/Entities/Modules/Settings/SerializationController.cs:2]: Another header just making note so we don't forget

File: DNN Platform/Library/Entities/Modules/Settings/SettingsRepository.cs [hunk 24]
```diff
@@ -2,20 +2,15 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT License. See LICENSE file in the project root for full license information.
```
Comment (mitchelsellers) [inline DNN Platform/Library/Entities/Modules/Settings/SettingsRepository.cs:3]: Another header just making note so we don't forget

**Comment by donker** [comment]
> 
> 
> I've done a partial review on this, my concern is that this PR seems to be including at least three distinctly different changes.
> 
>     * Moving of Prompt to the core libraries
> 
>     * Changing the way that settings for Modules are stored/serialized
> 
>     * Changes to assembly attributes impacting com visibility etc
> 
> 
> I almost want to have this as separate PRs

Do you know of a way to pry this apart? This stuff relies on each other, which is why I did this. The CLS compliance I did in the solutioninfo instead of all the various subprojects because we wanted to do this globally. But again, this is one commit. So not sure how to pry it out of the PR.

**Comment by mitchelsellers** [comment]
We might not be able to do that now, but I don't have an easy way to verify that we don't chang behavior as it is right now.

@bdukes Thoughts?

**Comment by donker** [comment]
The only thing I can think of is to diff the end result with a fresh develop checkout and then recreate the changes. Not something I'm really looking forward to, but if it has to be done I'll do it.

**Comment by donker** [comment]
OK, I'm going to do it the way I described above. Closing this as I'll replace it with separated PRs.

**Comment by valadas** [comment]
I am a bit late to the party, but if it is on it's own commit, you could create a new branch from develop and cherrypick individual commits from this PR branch.
